### PR TITLE
Just me spaces support

### LINKED ISSUE #3214: Space | Just me space first iteration
Modify wizard to include add existing rooms or open it post creation?

1d

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #3214

Added Just me space support.
After creation the add room screen will be open automatically.

Also contains update on Add room screen to separate spaces and room as per design

Edit: Added action to create a new room

image

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt [hunk 1]
```diff
@@ -42,8 +42,9 @@ import org.matrix.android.sdk.api.session.room.failure.CreateRoomFailure
 import org.matrix.android.sdk.api.session.room.model.RoomDirectoryVisibility
 import org.matrix.android.sdk.api.session.room.model.create.CreateRoomParams
 import org.matrix.android.sdk.api.session.room.model.create.CreateRoomPreset
+import timber.log.Timber
 
-class CreateRoomViewModel @AssistedInject constructor(@Assisted initialState: CreateRoomViewState,
+class CreateRoomViewModel @AssistedInject constructor(@Assisted val initialState: CreateRoomViewState,
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:47]: `initialState` could be private (I think)

File: vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt [hunk 2]
```diff
@@ -228,9 +230,21 @@ class CreateRoomViewModel @AssistedInject constructor(@Assisted initialState: Cr
...
         viewModelScope.launch {
-            val result = runCatching { session.createRoom(createRoomParams) }
-            result.fold(
+            runCatching { session.createRoom(createRoomParams) }.fold(
                     { roomId ->
+
+                        if (initialState.parentSpaceId != null) {
+                            // add it as a child
+                            try {
+                                val via = session.sessionParams.homeServerHost?.let { listOf(it) } ?: emptyList()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]:
```suggestion
                                val via = session.sessionParams.homeServerHost?.toList().orEmpty()
```

File: vector/src/main/java/im/vector/app/features/spaces/SpaceCreationActivity.kt [hunk 3]
```diff
@@ -106,6 +107,9 @@ class SpaceCreationActivity : SimpleFragmentActivity(), CreateSpaceViewModel.Fac
                     setResult(RESULT_OK, Intent().apply {
                         putExtra(RESULT_DATA_CREATED_SPACE_ID, it.spaceId)
                         putExtra(RESULT_DATA_DEFAULT_ROOM_ID, it.defaultRoomId)
+                        if (it.topology == SpaceTopology.JustMe) {
+                            putExtra(RESULT_DATA_CREATED_SPACE_IS_JUST_ME, true)
+                        }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/SpaceCreationActivity.kt:112]: I prefer to always put the extra value, even when value is false.

File: vector/src/main/java/im/vector/app/features/spaces/create/CreateSpaceViewModel.kt [hunk 4]
```diff
@@ -126,10 +126,12 @@ class CreateSpaceViewModel @AssistedInject constructor(
             SpaceTopology.JustMe -> {
                 setState {
                     copy(
-                            spaceTopology = SpaceTopology.JustMe
+                            spaceTopology = SpaceTopology.JustMe,
+                            defaultRooms = emptyMap()
                     )
                 }
                 // XXX finish and open the add rooms directly
+                handleNextFromDefaultRooms()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/create/CreateSpaceViewModel.kt:134]: `XXX` can be removed fro comment (it's like a TODO, no?)

**Review by bmarty — COMMENTED** [review]
A few small remarks, else LGTM

File: vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:47]: done

File: vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: done

File: vector/src/main/java/im/vector/app/features/spaces/SpaceCreationActivity.kt (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/SpaceCreationActivity.kt:112]: done

File: vector/src/main/java/im/vector/app/features/spaces/create/CreateSpaceViewModel.kt (hunk 4 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/create/CreateSpaceViewModel.kt:134]: fixed, thx

File: vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt [hunk 5]
```diff
@@ -236,7 +236,7 @@ class CreateRoomViewModel @AssistedInject constructor(@Assisted val initialState
                         if (initialState.parentSpaceId != null) {
                             // add it as a child
                             try {
-                                val via = session.sessionParams.homeServerHost?.let { listOf(it) } ?: emptyList()
+                                val via = session.sessionParams.homeServerHost?.let { listOf(it) }.orEmpty()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: there was another change in my review: use `toList()`. Is it possible?

Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: :D

Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: Oh my bad, this method really sucks for Strings

Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: maybe create an extension `fun String.asList() = listOf(this)` :)

Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/roomdirectory/createroom/CreateRoomViewModel.kt:239]: TBH, i think i will refactor this soon and see how to reuse the permalink service computeViaParams

**Review by bmarty — APPROVED** [review]
Thanks for the update!

**Pull request merged**